src/Microsoft.Diagnostics.Monitoring.EventPipe/EventPipeStreamProvider.cs

```diff
@@ -28,7 +30,44 @@ public async Task<Stream> ProcessEvents(DiagnosticsClient client, TimeSpan durat
             EventPipeSession session = null;
             try
             {
-                session = await client.StartEventPipeSessionAsync(_sourceConfig.GetProviders(), _sourceConfig.RequestRundown, _sourceConfig.BufferSizeInMB, cancellationToken).ConfigureAwait(false);
+                IEnumerable<EventPipeProvider> providers = _sourceConfig.GetProviders();
+                int bufferSizeInMB = _sourceConfig.BufferSizeInMB;
+                long rundownKeyword = _sourceConfig.RundownKeyword;
+                RetryStrategy retryStrategy = _sourceConfig.RetryStrategy;
+                bool retry = true;
+                while (retry)
+                {
+                    retry = false;
+                    try
+                    {
+                        EventPipeSessionConfiguration config = new(providers, bufferSizeInMB, (rundownKeyword != 0), true, rundownKeyword);
+                        session = await client.StartEventPipeSessionAsync(config, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (UnsupportedCommandException e)
+                    {
+                        if (retryStrategy == RetryStrategy.DropKeywordKeepRundown)
+                        {
+                            Debug.Assert(rundownKeyword != 0);
+                            Debug.Assert(rundownKeyword != EventPipeSession.DefaultRundownKeyword);
+                            retry = true;
+                            retryStrategy = RetryStrategy.DoNotRetry;
+                            rundownKeyword = 0;
+                        }
+                        else if (retryStrategy == RetryStrategy.DropKeywordDropRundown)
+                        {
+                            Debug.Assert(rundownKeyword != 0);
+                            Debug.Assert(rundownKeyword != EventPipeSession.DefaultRundownKeyword);
```

Comment:
Same here, CollectTracing2 may have been what was unsupported.

Reply:
I am not sure I understand this one - it appears the upstream is providing invalid combination for the code to reach here and the intention is to catch that.

If the combined option implies we should be dropping the keyword and dropping the rundown in the next retry, then the request we sent earlier should have a custom rundown keyword to begin with.

Suppose the actual targeted runtime does not support `CollectTracing2`, in the first round, we would have sent a request trying with a custom rundown keyword with the `RetryStrategy` set to `DropKeywordDropRundown`, reaching this branch, and the assert is okay at this point, then we set the RetryStrategy to `DoNotRetry`, and then retry without the custom keyword, and the next time around, the code will not reach this branch, which mean the assertion is still correct? 

The only case this could be a problem is somehow the upstream is not specifying a custom rundown keyword but then specified the `RetryStrategy` to `DropKeywordDropRundown`, which is what this assert is meant to catch, the upstream should not be doing that.

Reply:
> The only case this could be a problem is somehow the upstream is not specifying a custom rundown keyword but then specified the RetryStrategy to DropKeywordDropRundown, which is what this assert is meant to catch, the upstream should not be doing that, it will result in useless retries.

It sounds like you are assuming that a MonitoringSource with Rundown=0 and RetryStrategy=DropKeywordDropRundown is a bad configuration. I see it as being useful. Imagine we have two different configs:
SourceConfig1 { RundownKeyword = 0, RetryStrategy = DoNotRetry }
SourceConfig2 { RundownKeyword = GC, RetryStrategy = DropKeywordDropRundown }

If I combine those two into an aggregate I will get { RundownKeyword = GC, RetryStrategy = DoNotRetry }. Ideally I would like the GC keyword to still fallback but because SourceConfig said NoRetry it no longer does. However if SourceConfig1 had been specified as { RundownKeyword = 0, RetryStrategy = DropKeywordDropRundown } then the aggregate would also have Retry=DropKeywordDropRundown and it would have the fallback behavior we want. That is the scenario where { RundownKeyword = 0, RetryStrategy = DropKeywordDropRundown } appears very useful.

Ultimately I think the difference is how DoNotRetry is being interpreted. One interpretation is "this config doesn't need a retry, but its allowable". The other interpretation is "This config is not allowed to be retried". Currently I was thinking of DoNotRetry as the latter interpretation and that is also what the override rules currently implement. We could define an enumeration value to mean "Not needed but allowable", however I don't think we need to go out of our way to optimize away one unneeded retry. As best I can tell doing one extra retry has no downside other than a neglible perf cost.

Thoughts?

Reply:
As is of the 2nd iteration, the `DoNotRetry` notion allows multiple interpretations, and this could be a source of bugs moving forward. 

I think I value clarity of the code more. Saving one retry for performance was not the goal, the goal is to make it clear what does the enum value really means.

In the new iteration, I split the concept of `DoNotRetry` into `NothingToRetry` and `ForbiddenToRetry`

`NothingToRetry` means literally that, there is nothing we could do to save it - in the merge case, if something else got something retriable, then we can always override.

`ForbiddenToRetry` always means literally that, as an author, I forbid any retry activities. I am using newer features, but it is critical, retry with that newer feature dropped means nothing to the user, might as well not retry at all.

With that, your scenario turn into something like this:

```
SourceConfig1 { RundownKeyword = 0, RetryStrategy = NothingToRetry }
SourceConfig2 { RundownKeyword = GC, RetryStrategy = DropKeywordDropRundown }
```

And so the merged config is this:


```
Merged { RundownKeyword = GC, RetryStrategy = DropKeywordDropRundown }
```

And it will work just fine in both the retry or not retry case.

On the other hand, you can also do:

```
SourceConfig1 { RundownKeyword = 0, RetryStrategy = ForbiddenToRetry }
SourceConfig2 { RundownKeyword = GC, RetryStrategy = DropKeywordDropRundown }
```

And the merged config will be this:

```
Merged { RundownKeyword = GC, RetryStrategy = ForbiddenToRetry }
```

And now we execute as instructed, do not retry.

Reply:
I think you could still achieve clarity without having a 'NothingToRetry' option, but I don't have a strong preference. If you want to have a 'NothingToRetry' value I won't object :)